content/blog/2023-12-11-bloke/index.md

```diff
@@ -0,0 +1,372 @@
++++
+title = "BLOKE: Optimizing Bril with STOKE"
+[extra]
+bio = """
+    Kei Imada is a first-year Ph.D. student. He is interested in applying
+    mathematical structures to type systems and formal verification to
+    enable efficient development of parallel and distributed systems, and software
+    defined networks.
+"""
+[[extra.authors]]
+name = "Kei Imada"
+link = "https://keikun555.github.io/"
++++
+
+# BLOKE: Optimizing BRIL with STOKE 
+[Presentation](assets/presentation.pdf)
+[Report](assets/report.pdf)
+
+## Abstract
+
+We present bloke, a scalable implementation of STOKE on an educational intermediate representation language. In contrast to STOKE's two-stage process, BLOKE supports a user-defined number of phases that smoothly transitions from synthesis to optimization. We found that BLOKE was able to find superoptimizations of loop-free code written in Bril. Furthermore, we found BLOKE to be scalable with respect to the number of cores a machine has.
+
+## Introduction
+
+Compiler optimizations usually seek to improve a metric on a program. Dead code elimination decreases the number of instructions of programs and loop invariant code motion aims to decrease the number of instructions executed within a loop.
+
+Contrary to the name "optimizations," these algorithms don't seek to optimize, or find the best possible program for the metric. Given this misnomer, Alexia Massalin coined the term superoptimizers.
+
+Given an instruction set, the superoptimizer finds the shortest program to compute a function. [3]
+
+Most problems within the field of superoptimizations are intractable, and therefore we generally employ exhaustive bottom-up search or stochastic search.
+
+In this paper, we apply stochastic methods to find superoptimizations of loop-free programs. Similar to STOKE [8], we use MarkovChain Monte-Carlo (MCMC) methods to conduct a random walk on the space of programs.
+
+In contrast with STOKE, which uses a twostage process to introduce the program performance component, our BLOKE implementation gradually introduces the performance component into the MCMC search.
+We will first give a primer on Bril, the Metropolis Hastings algorithm, and STOKE, the precursor to BLOKE. We then describe BLOKE, our stochastic superoptimizer for loop-free Bril programs. Finally, we evaluate BLOKE to show that BLOKE can find superoptimizations of Bril programs scalably.
+
+## Bril
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-01.jpg?height=673&width=569&top_left_y=1342&top_left_x=1189)
```

Comment:
This image seems to just contain code. Consider instead just including it inline as text here, in a Markdown code block.